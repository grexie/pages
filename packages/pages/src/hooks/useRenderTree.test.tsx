import { act, renderHook } from '@testing-library/react';
import {
  createContext,
  Dispatch,
  PropsWithChildren,
  SetStateAction,
  useContext,
  useLayoutEffect,
  useState,
  FC,
} from 'react';
import {
  RenderTreeNode,
  useRenderTreeNode,
  withRenderTree,
} from './useRenderTree.js';
import { expect } from 'chai';
import * as chai from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';

chai.use(jestSnapshotPlugin());

describe('Grexie Pages', () => {
  describe('useRenderTree', () => {
    it('test', async () => {
      const { result } = renderHook(() => useRenderTreeNode());
      expect(result.current).to.be.undefined;
    });

    it('test2', async () => {
      const MutateContext = createContext<Dispatch<SetStateAction<number>>>(
        null as any
      );

      const TreeNode = withRenderTree(({ children }: PropsWithChildren) => (
        <>{children}</>
      ));

      const { result, rerender } = renderHook(
        ({ mutate }: { mutate: number }) => {
          const setMutate = useContext(MutateContext);
          useLayoutEffect(() => {
            setMutate(mutate);
          }, [mutate]);
          return useRenderTreeNode();
        },
        {
          wrapper: (({ children }) => {
            const [mutate, setMutate] = useState(0);

            return (
              <MutateContext.Provider value={setMutate}>
                <TreeNode>
                  <TreeNode />
                  <TreeNode>
                    <TreeNode />
                    <TreeNode />
                    {mutate >= 2 && (
                      <TreeNode>
                        <TreeNode>
                          <TreeNode />
                        </TreeNode>
                      </TreeNode>
                    )}
                    {mutate >= 1 && (
                      <TreeNode>
                        <TreeNode />
                      </TreeNode>
                    )}
                    <TreeNode>{children}</TreeNode>
                  </TreeNode>
                  <TreeNode />
                </TreeNode>
              </MutateContext.Provider>
            );
          }) as FC<PropsWithChildren>,
          initialProps: {
            mutate: 0,
          },
        }
      );
      expect(result.current).instanceOf(RenderTreeNode);
      expect(result.current.root).instanceOf(RenderTreeNode);
      expect(result.current.root).not.equal(result.current);
      expect(result.current.root).toMatchSnapshot('mutate-0');
      act(() => {
        rerender({ mutate: 1 });
      });
      expect(result.current.root).toMatchSnapshot('mutate-1');
      act(() => {
        rerender({ mutate: 2 });
      });
      expect(result.current.root).toMatchSnapshot('mutate-2');
      act(() => {
        rerender({ mutate: 1 });
      });
      expect(result.current.root).toMatchSnapshot('mutate-1');
      act(() => {
        rerender({ mutate: 2 });
      });
      expect(result.current.root).toMatchSnapshot('mutate-2');
    });
  });
});
