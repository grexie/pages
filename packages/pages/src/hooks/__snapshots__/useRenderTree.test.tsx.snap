// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grexie Pages useRenderTree test2 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [],
          "index": 2,
          "order": 5,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 6,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`Grexie Pages useRenderTree test2 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [],
          "index": 3,
          "order": 7,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 8,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`Grexie Pages useRenderTree test2 3`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "index": 0,
                  "order": 7,
                },
              ],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 9,
            },
          ],
          "index": 3,
          "order": 8,
        },
        Object {
          "children": Array [],
          "index": 4,
          "order": 10,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 11,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`Grexie Pages useRenderTree test2 4`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [],
          "index": 3,
          "order": 7,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 8,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`Grexie Pages useRenderTree test2 5`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "index": 0,
                  "order": 7,
                },
              ],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 9,
            },
          ],
          "index": 3,
          "order": 8,
        },
        Object {
          "children": Array [],
          "index": 4,
          "order": 10,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 11,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`useRenderTree test2 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [],
          "index": 2,
          "order": 5,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 6,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`useRenderTree test2 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [],
          "index": 3,
          "order": 7,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 8,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`useRenderTree test2 3`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "index": 0,
                  "order": 7,
                },
              ],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 9,
            },
          ],
          "index": 3,
          "order": 8,
        },
        Object {
          "children": Array [],
          "index": 4,
          "order": 10,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 11,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`useRenderTree test2 4`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [],
          "index": 3,
          "order": 7,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 8,
    },
  ],
  "index": 0,
  "order": 0,
}
`;

exports[`useRenderTree test2 5`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "index": 0,
      "order": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "index": 0,
          "order": 3,
        },
        Object {
          "children": Array [],
          "index": 1,
          "order": 4,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "index": 0,
                  "order": 7,
                },
              ],
              "index": 0,
              "order": 6,
            },
          ],
          "index": 2,
          "order": 5,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "index": 0,
              "order": 9,
            },
          ],
          "index": 3,
          "order": 8,
        },
        Object {
          "children": Array [],
          "index": 4,
          "order": 10,
        },
      ],
      "index": 1,
      "order": 2,
    },
    Object {
      "children": Array [],
      "index": 2,
      "order": 11,
    },
  ],
  "index": 0,
  "order": 0,
}
`;
